GUIDELINES.md

```diff
@@ -0,0 +1,892 @@
+
+1. The **test class** should define **WHAT** is tested (the "login" feature, test cases, ...)
+2. The **steps class** should define **HOW** it is tested (technically, which classes are involved, where we put mocks, ...)
+
+### Define test cases
+
+So let's add a test case (see [final class](https://github.com/mysugr/sweetest/blob/example-upgrade/app/src/test/java/com/mysugr/android/testing/example/view/LoginTest.kt) ):
+
+```kotlin
+@Test
+fun `Login with correct credentials is successful`() {
+    sut {
+        givenExistingUser(email = EXISTING_EMAIL, password = EXISTING_PASSWORD, authToken = EXISTING_AUTH_TOKEN)
+        whenLoggingIn(email = EXISTING_EMAIL, password = EXISTING_PASSWORD)
+        thenEmailWasCheckedAtBackend(EXISTING_EMAIL)
+    }
+}
+
+companion object {
+    const val EXISTING_EMAIL = "existing@test.com"
+    const val EXISTING_PASSWORD = "supersecure1"
+    const val EXISTING_AUTH_TOKEN = "auth_token"
+}
+```
+
+`sut { ... }` is used in order to get inside the scope of the steps class and call some functions there. As you can see the test is quite expressive and could as well be read by non-technical people. That's exactly as it should be. All the technical implementation is happening in the steps class then.
+
+**Tip:** you can write all your test cases like that in the test class first and then let the IDE create the missing functions in the respective steps class (e.g. in IntelliJ: Option + Enter, select "Create member function" and select the target class, in this case the steps class). That way you can flesh out the tests TDD-style without needing to care about the technical implementation in the steps class yet.
+
+### Add configuration in the steps class
+
+In order to know how we set up the test we should first quickly get a grasp of the system we want to test. Here is an outline of the example with a typical Android architecture:
+
+```
+┏━━━━━━━━━━━━━━━┓
+┃ LoginActivity ┃  ⟵ Android view
+┗━━━━━━━━━━━━━━━┛
+
+━━━ ᐁ UNDER TEST ᐁ ━━━
+
+┏━━━━━━━━━━━━━━━━┓
+┃ LoginViewModel ┃  ⟵ presentation
+┗━━━━━━━━━━━━━━━━┛
+┏━━━━━━━━━━━━━┓ 
+┃ AuthManager ┃  ⟵ domain
+┗━━━━━━━━━━━━━┛
+┏━━━━━━━━━━━━━━━━━┓ ┏━━━━━━━━━━━━━━━┓
+┃ BackendGateway* ┃ ┃ SessionStore* ┃  ⟵ data access
+┗━━━━━━━━━━━━━━━━━┛ ┗━━━━━━━━━━━━━━━┛
+* = mocked
+```
+
+In order to achieve that you have to add a configuration that reflects the wanted setup in the steps file:
+
+```kotlin
+override fun configure() = super.configure()
+    .requireReal<LoginViewModel>()
+    .requireReal<AuthManager>()
+```
+
+This would be already enough in terms of configuration to get the dependency graph created automatically! sweetest will provide Mockito mocks automatically if the dependency isn't configured otherwise, that means that `LoginViewModel` and `AuthManager` will be instantiated and the rest of the dependencies (`BackendGateway` and `SessionStore`) will be Mockito mocks.
```

Comment:
`any` just means that the dependency has the chance to become both a mock and a real instance. but it can be determined at test run time.

so no, it works also if you e.g. use `mockOnly` or `realOnly`, then you don't need to use `requireMock` or `requireReal` in steps/test classes.

does this answer your question? i'm not 100% sure i understood your question correctly.

Reply:
So when I don't have `any` in the configuration, but `requireReal`, it runs?  I didn't get to reference yet, but a table there with what works and what not would be nice :)

Reply:
i again don't understand your question 🤣 what do you mean by "don't have `any`"? when you have `any` in the module config it's the least constraining you can have. so `requireReal` down in the steps/class file would work. just if you set `mockOnly` in the module config and `requireReal` it's a conflicting configuration and then it fails.

i don't know why we should add a comparison table here. first, tha place you tagged is probably the wrong one her, and second, dependency configuration on a per-module basis will be deprecated. but maybe i didn't understand fully what you mean (still 😁), if you want we can have a call 😉

Reply:
I just remember my sweetest journey and that it is easy to forget adding the dependency in the moduleConfiguration. So a sentence like: "Also make sure the dependency you add here with `requireReal` has to be defined in the module config, see above." 

A table could be added in Reference I meant. Here it doesn't make sense of course.

> second, dependency configuration on a per-module basis will be deprecated.
That was actually my question, whether this will be improved. Nice to hear that 👍

Reply:
> I just remember my sweetest journey and that it is easy to forget adding the dependency in the moduleConfiguration. So a sentence like: "Also make sure the dependency you add here with `requireReal` has to be defined in the module config, see above."

makes total sense 👍 

> A table could be added in Reference I meant. Here it doesn't make sense of course.

you mean something like this?

| Configuration | Module testing configuration | allowed? |
|:---|:--|:--|
| `requireMock` | `any` | ✅ (`any` doesn't constraint mode) |
| `requireMock` | `realOnly` | ❌ (`realOnly` doesn't allow mocks) |
| `requireReal` | `realOnly` | ✅ (`realOnly` allows `requireReal`) |
| `requireReal` | `mockOnly` | ❌ (`mockOnly` doesn't allow `requireReal`)|
| `requireReal` | `any` | ✅ (`any` doesn't constraint mode) |

i think that's redundant somehow...

Reply:
Haha, yes exactly. Nice that you did it and then say it is redundant. I would have order the columns differently, module testing config first and then Steps Configuration. There would be even more situations (e.g. realOnly and no config in steps etc). Instead of allowed, the last column would be: It is mock or real or not allowed. I would like a table like that as reference a lot, but just my opinion

Reply:
ok, for time's sake i moved it to my notes so it's respected in the next update, which will anyways happen soon because of the mockk-update. reason: i need to wrap my head around the upcoming deprecation of the dependency configuration in the module testing configuration. i hope that's ok for you!